include/sunlinsol/sunlinsol_ginkgo.hpp

```diff
@@ -0,0 +1,348 @@
+/// This checks if:
+/// 1. Was the absolute residual tolerance met?
+/// 2. Was the max iteration count reached?
+class DefaultStop : public gko::EnablePolymorphicObject<DefaultStop, gko::stop::Criterion>
+{
+  friend class gko::EnablePolymorphicObject<DefaultStop, gko::stop::Criterion>;
+
+public:
+  GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+  {
+    sunrealtype GKO_FACTORY_PARAMETER_SCALAR(reduction_factor,
+                                             SUN_UNIT_ROUNDOFF); // NOLINT(cppcoreguidelines-avoid-magic-numbers)
+    gko::uint64 GKO_FACTORY_PARAMETER_SCALAR(max_iters, 5);
+  };
+  GKO_ENABLE_CRITERION_FACTORY(DefaultStop, parameters, Factory);
+  GKO_ENABLE_BUILD_METHOD(Factory);
+
+  gko::uint64 get_max_iters() const
+  {
+    return parameters_.max_iters;
+  }
+
+  sunrealtype get_reduction_factor() const
+  {
+    return parameters_.reduction_factor;
+  }
+
+protected:
+  bool check_impl(gko::uint8 stoppingId, bool setFinalized, gko::array<gko::stopping_status>* stop_status,
+                  bool* one_changed, const Updater&) override;
+
+  explicit DefaultStop(std::shared_ptr<const gko::Executor> exec)
+      : EnablePolymorphicObject<DefaultStop, gko::stop::Criterion>(std::move(exec))
+  {}
+
+  explicit DefaultStop(const Factory* factory, const gko::stop::CriterionArgs& args)
+      : EnablePolymorphicObject<DefaultStop, gko::stop::Criterion>(factory->get_executor()), parameters_{
+                                                                                                 factory->get_parameters()}
+  {
+    criteria_.push_back(gko::stop::ResidualNorm<sunrealtype>::build()
+                            .with_reduction_factor(parameters_.reduction_factor)
+                            .with_baseline(gko::stop::mode::absolute)
+                            .on(factory->get_executor())
+                            ->generate(args));
+    criteria_.push_back(
+        gko::stop::Iteration::build().with_max_iters(parameters_.max_iters).on(factory->get_executor())->generate(args));
+  }
+
+private:
+  std::vector<std::unique_ptr<Criterion>> criteria_{};
+};
+
+inline bool DefaultStop::check_impl(gko::uint8 stoppingId, bool setFinalized,
+                                    gko::array<gko::stopping_status>* stop_status, bool* one_changed,
+                                    const Updater& updater)
+{
+  bool one_converged = false;
+  gko::uint8 ids{1};
+  *one_changed = false;
+  for (auto& c : criteria_) {
+    bool local_one_changed = false;
+    one_converged |= c->check(ids, setFinalized, stop_status, &local_one_changed, updater);
+    *one_changed |= local_one_changed;
+    if (one_converged) {
+      break;
+    }
+    ids++;
+  }
+  return one_converged;
+}
+
+/// Class that wraps a Ginkgo solver (factory) and is convertible to a fully
+/// functioning ``SUNLinearSolver``.
+template<class GkoSolverType, class GkoMatrixType>
+class LinearSolver : public ConvertibleTo<SUNLinearSolver>
```

Comment:
In the batched branch (if it's not done already) we should update this to use `sundials::impl::BaseLinearSolver`,